server/src/main/java/org/elasticsearch/plugins/ModuleSupport.java

```diff
@@ -82,31 +88,44 @@ static ModuleDescriptor createModuleDescriptor(
         Map<String, List<String>> allBundledProviders = new HashMap<>();
         Set<String> servicesUsedInBundle = new HashSet<>();
         for (Path path : jarPaths) {
+            logger.info("Scanning JAR " + path + " for services");
             assert path.getFileName().toString().endsWith(".jar") : "expected jars suffix, in path: " + path;
             try (JarFile jf = new JarFile(path.toFile(), true, ZipFile.OPEN_READ, Runtime.version())) {
                 // if we have a module declaration, trust its uses/provides
                 JarEntry moduleInfo = jf.getJarEntry("module-info.class");
                 if (moduleInfo != null) {
                     var descriptor = getDescriptorForModularJar(path);
-                    pkgs.addAll(descriptor.packages());
+                    pkgs.addAll(descriptor.packages().stream().filter(Predicate.not(FORBIDDEN_PACKAGE_NAMES::contains)).toList());
                     servicesUsedInBundle.addAll(descriptor.uses());
+                    logger.info("Adding modular uses " + String.join(";", descriptor.uses()));
+
                     for (ModuleDescriptor.Provides p : descriptor.provides()) {
                         String serviceName = p.service();
                         List<String> providersInModule = p.providers();
+                        logger.info("Adding modular providers " + String.join(";", providersInModule));
 
                         allBundledProviders.compute(serviceName, (k, v) -> createListOrAppend(v, providersInModule));
                         servicesUsedInBundle.add(serviceName);
                     }
                 } else {
                     var scan = scan(jf);
-                    scan.classFiles().stream().map(cf -> toPackageName(cf, "/")).flatMap(Optional::stream).forEach(pkgs::add);
+                    scan.classFiles()
+                        .stream()
+                        .map(cf -> toPackageName(cf, "/"))
+                        .flatMap(Optional::stream)
+                        .filter(Predicate.not(FORBIDDEN_PACKAGE_NAMES::contains))
+                        .forEach(pkgs::add);
 
                     // read providers from the list of service files
                     for (String serviceFileName : scan.serviceFiles()) {
                         String serviceName = getServiceName(serviceFileName);
                         List<String> providersInJar = getProvidersFromServiceFile(jf, serviceFileName);
 
-                        allBundledProviders.compute(serviceName, (k, v) -> createListOrAppend(v, providersInJar));
+                        if (providersInJar.isEmpty() == false) {
+                            logger.info("Adding non-modular providers " + String.join(";", providersInJar));
```

Comment:
keep in mind this is WIP, very "debugging" PR - these `info` logs will become debug or be removed altogether, (if they stay, they'll probably become the lambda accepting kind, to avoid computing the String.join if not needed)